targeting: Add device tree based targeting support

devenrao opened the pull request:
This commit introduces a new targeting infrastructure that builds on a flattened device tree (FDT) to construct target objects and resolve hierarchical associations like physical and affinity paths. Key features:

- `TargetService`: Singleton service that loads and indexes targets from a DTB file using `TargetDevtreeMap`.
- Support for `ATTR_PHYS_DEV_PATH`, entity path parsing, and target associations via `parentByPhysical` and `parentByAffinity`.
- `PredicateAttrVal` and `PredicatePostfixExpr` support to compose flexible attribute-based queries on targets.

Tests added for:
- Enumerating targets by physical and affinity associations
- Round-tripping between EntityPath and Target
- Attribute-based predicates using AND, OR, and NOT logic
- Immediate parent and child target retrieval

Test1: All targets with ATTR_PHYS_DEV_PATH childByPhysical all
  physical:sys-0/node-0
  physical:sys-0/node-0/proc-0
  physical:sys-0/node-0/proc-0/mc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0/omi-0
  physical:sys-0/node-0/ocmb-1
Test2: All targets with ATTR_PHYS_DEV_PATH childByAffinity all
  physical:sys-0/node-0
  physical:sys-0/node-0/proc-0
  physical:sys-0/node-0/proc-0/mc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0/omi-0
  physical:sys-0/node-0/ocmb-0
Test3: PredicatePostFoxExpr AttrVal and Attr
PROC: physical:sys-0/node-0/proc-0
  └── physical:sys-0/node-0/proc-0/mc-0
PROC: physical:sys-0/node-0/proc-1
  └── physical:sys-0/node-0/proc-1/mc-0
Convert ocmb1 binary data to EntityPath and to Target
 entitypath to ocmb physical path physical:sys-0/node-0/ocmb-1
 parent of ocmb parent affinity path physical:sys-0/node-0
 parent of ocmb parent physical path physical:sys-0/node-0/proc-1/mc-0/mi-0/mcc-0/omi-0

[ RUN      ] TargetServiceTest.TestFDTNotNull
[       OK ] TargetServiceTest.TestFDTNotNull (0 ms)
[ RUN      ] TargetServiceTest.TestTopLevelTarget
[       OK ] TargetServiceTest.TestTopLevelTarget (0 ms)
[ RUN      ] TargetServiceTest.TestGetParentOfOcmbImmediatePhysical
[       OK ] TargetServiceTest.TestGetParentOfOcmbImmediatePhysical (0 ms)
[ RUN      ] TargetServiceTest.TestGetParentOfOcmbImmediateAffinity
[       OK ] TargetServiceTest.TestGetParentOfOcmbImmediateAffinity (0 ms)
[ RUN      ] TargetServiceTest.TestGetAssociatedChildrenImmediate
[       OK ] TargetServiceTest.TestGetAssociatedChildrenImmediate (0 ms)
[ RUN      ] TargetServiceTest.TestGetAssociatedParentsAffinityAll
[       OK ] TargetServiceTest.TestGetAssociatedParentsAffinityAll (0 ms)
[ RUN      ] TargetServiceTest.TestRoundTripToTarget
[       OK ] TargetServiceTest.TestRoundTripToTarget (0 ms)
[ RUN      ] TargetServiceTest.TestPredicateAttrValProcType
[       OK ] TargetServiceTest.TestPredicateAttrValProcType (0 ms)
[ RUN      ] TargetServiceTest.TestPredicatePostfixExpr_AttrVal_AND
came into for loop
came into for loop
[       OK ] TargetServiceTest.TestPredicatePostfixExpr_AttrVal_AND (0 ms)
[ RUN      ] TargetServiceTest.TestPredicatePostfixExpr_AttrMask_OR
[       OK ] TargetServiceTest.TestPredicatePostfixExpr_AttrMask_OR (0 ms)
[ RUN      ] TargetServiceTest.TestPredicatePostfixExpr_Negation
[       OK ] TargetServiceTest.TestPredicatePostfixExpr_Negation (0 ms)
[----------] 11 tests from TargetServiceTest (0 ms total)


Change-Id: I89c727daab165ba4ba9835c65769835b5c1d7b65

devenrao committed:
targeting: Add new DT-based TargetService support

This commit introduces a new targeting infrastructure that builds on a
flattened device tree (FDT) to construct target objects and resolve
hierarchical associations like physical and affinity paths. Key features:

- `TargetService`: Singleton service that loads and indexes targets
  from a DTB file using `TargetDevtreeMap`.
- Support for `ATTR_PHYS_DEV_PATH`, entity path parsing, and target
  associations via `parentByPhysical` and `parentByAffinity`.
- `PredicateAttrVal` and `PredicatePostfixExpr` support to compose
  flexible attribute-based queries on targets.

Tests added for:
- Enumerating targets by physical and affinity associations
- Round-tripping between EntityPath and Target
- Attribute-based predicates using AND, OR, and NOT logic
- Immediate parent and child target retrieval

Test1: All targets with ATTR_PHYS_DEV_PATH childByPhysical all
  physical:sys-0/node-0
  physical:sys-0/node-0/proc-0
  physical:sys-0/node-0/proc-0/mc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0/omi-0
  physical:sys-0/node-0/ocmb-1
Test2: All targets with ATTR_PHYS_DEV_PATH childByAffinity all
  physical:sys-0/node-0
  physical:sys-0/node-0/proc-0
  physical:sys-0/node-0/proc-0/mc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0
  physical:sys-0/node-0/proc-0/mc-0/mi-0/mcc-0/omi-0
  physical:sys-0/node-0/ocmb-0
Test3: PredicatePostFoxExpr AttrVal and Attr
PROC: physical:sys-0/node-0/proc-0
  └── physical:sys-0/node-0/proc-0/mc-0
PROC: physical:sys-0/node-0/proc-1
  └── physical:sys-0/node-0/proc-1/mc-0
Convert ocmb1 binary data to EntityPath and to Target
 entitypath to ocmb physical path physical:sys-0/node-0/ocmb-1
 parent of ocmb parent affinity path physical:sys-0/node-0
 parent of ocmb parent physical path physical:sys-0/node-0/proc-1/mc-0/mi-0/mcc-0/omi-0

[ RUN      ] TargetServiceTest.TestFDTNotNull
[       OK ] TargetServiceTest.TestFDTNotNull (0 ms)
[ RUN      ] TargetServiceTest.TestTopLevelTarget
[       OK ] TargetServiceTest.TestTopLevelTarget (0 ms)
[ RUN      ] TargetServiceTest.TestGetParentOfOcmbImmediatePhysical
[       OK ] TargetServiceTest.TestGetParentOfOcmbImmediatePhysical (0 ms)
[ RUN      ] TargetServiceTest.TestGetParentOfOcmbImmediateAffinity
[       OK ] TargetServiceTest.TestGetParentOfOcmbImmediateAffinity (0 ms)
[ RUN      ] TargetServiceTest.TestGetAssociatedChildrenImmediate
[       OK ] TargetServiceTest.TestGetAssociatedChildrenImmediate (0 ms)
[ RUN      ] TargetServiceTest.TestGetAssociatedParentsAffinityAll
[       OK ] TargetServiceTest.TestGetAssociatedParentsAffinityAll (0 ms)
[ RUN      ] TargetServiceTest.TestRoundTripToTarget
[       OK ] TargetServiceTest.TestRoundTripToTarget (0 ms)
[ RUN      ] TargetServiceTest.TestPredicateAttrValProcType
[       OK ] TargetServiceTest.TestPredicateAttrValProcType (0 ms)
[ RUN      ] TargetServiceTest.TestPredicatePostfixExpr_AttrVal_AND
came into for loop
came into for loop
[       OK ] TargetServiceTest.TestPredicatePostfixExpr_AttrVal_AND (0 ms)
[ RUN      ] TargetServiceTest.TestPredicatePostfixExpr_AttrMask_OR
[       OK ] TargetServiceTest.TestPredicatePostfixExpr_AttrMask_OR (0 ms)
[ RUN      ] TargetServiceTest.TestPredicatePostfixExpr_Negation
[       OK ] TargetServiceTest.TestPredicatePostfixExpr_Negation (0 ms)
[----------] 11 tests from TargetServiceTest (0 ms total)

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I89c727daab165ba4ba9835c65769835b5c1d7b65

dhruvibm commented on target/targeting/target.H:
```diff
@@ -0,0 +1,104 @@
+class Target;
+using TargetPtr = std::unique_ptr<Target>;
+using TargetPtrList = std::vector<TargetPtr>;
+class Target
```
getAttr and setAttr are missing?

devenrao commented on target/libfdt/.gitignore:
```diff
@@ -0,0 +1 @@
+libfdt.so.1
```
no need to review libfdt taken latest release and added to project

gcpin commented on target/targeting/dtree_loader.C:
```diff
@@ -0,0 +1,58 @@
+{
+    if (_fdt != nullptr)
+    {
+        munmap(_fdt, _size);
```
munmap can fail too similar to mmap , runtime_error  can be thrown here also.

gcpin commented on target/targeting/xmltohb/attributestructs.H:
```diff
@@ -0,0 +1,1149 @@
+
+#ifndef TARG_ATTRIBUTESTRUCTS_H
+#define TARG_ATTRIBUTESTRUCTS_H
```
other files are using pragma directive, any reason to use macros for this case?

gcpin commented on target/targeting/xmltohb/attributestructs.H:
```diff
@@ -0,0 +1,1149 @@
+
+namespace TARGETING
+{
+#define PACKED __attribute__((packed))
```
PACKED is already defined in builtins.h, isn't this redefinition ?

gcpin commented on target/targeting/common/entitypath.H:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        auto operator<=>(const PathElement&) const = default;
+    } __attribute__((packed));
```
Can we use PACKED defined in builtins.h instead ?

gopichand-paturi commented on target/targeting/common/entitypath.H:
```diff
@@ -0,0 +1,106 @@
+    EntityPath() = default;
+    explicit EntityPath(PathType type);
+
+    EntityPath copyRemoveLast() const;
```
If we have Move constructor, I think this copy could be more efficient

deepakala-k commented on target/targeting/target_service.H:
```diff
@@ -0,0 +1,58 @@
+    TargetService(const TargetService&) = delete;
+    TargetService& operator=(const TargetService&) = delete;
+
+    [[nodiscard]] bool isInitialized() const noexcept
```
Suggestion: Does it really need to nodiscard? All of the times, when we use is<Method> we check its return. This could be used for other methods like initialize, etc, where we fail to check the result.

deepakala-k commented on target/targeting/common/entitypath.H:
```diff
@@ -0,0 +1,106 @@
+    // Get path type (upper nibble)
+    PathType type() const
+    {
+        return static_cast<PathType>((_header >> 4) & 0x0F);
```
When adding a new type of target, it may not be part of enum PathType. Can we get this from xmltohb? And the static_cast might not check if it is a valid enum. Do we need to add a validate method?

deepakala-k commented on target/targeting/common/entitypath.H:
```diff
@@ -0,0 +1,106 @@
+    // Set size (lower nibble)
+    void setSize(uint8_t size)
+    {
+        _header = (_header & 0xF0) | (size & 0x0F);
```
0x0F is used in multiple places. Can we call it LOWER_NIBBLE/ HIGHER_NIBBLE. Might be difficult to catch errors

deepakala-k commented on target/targeting/predicates/predicatepostfixexpr.C:
```diff
@@ -0,0 +1,116 @@
+                    bool r = evalStackItem(top, target);
+                    top = !r;
+                    break;
+                }
```
case LogicalOp::Valid not handled?

deepakala-k commented on target/targeting/target_devtree_map.C:
```diff
@@ -0,0 +1,191 @@
+{
+    if (!child)
+    {
+        return nullptr;
```
there are few places where we return nullptr without proper traces, which was one of our painpoints, not having enough traces. can we do somethg like 

template<typename T>
T logAndReturnNull(const std::string& reason)
{
    std::cerr << "Returning error due to " << reason << std::endl;
    return nullptr;
}

#define LOG_AND_RETURN_NULL(reason) logAndReturnNull<decltype(nullptr)>(reason, __FILE__, __LINE__)

//And call like this 
return LOG_AND_RETURN_NULL("The passed child  pointer is invalid ");

deepakala-k commented on target/targeting/common/entitypath.C:
```diff
@@ -0,0 +1,188 @@
+        case TYPE_NODE:
+            return "Node";
+        case TYPE_PROC:
+        case TYPE_MI:
```
TYPE_MI has the wrong return. Please fix

devenrao commented on target/targeting/xmltohb/attributeenums.H:
```diff
@@ -0,0 +1,199 @@
+#pragma once
```
auto-generated

devenrao commented on target/targeting/xmltohb/attributetraits.H:
```diff
@@ -0,0 +1,323 @@
+#pragma once
```
auto-generated

gcpin commented on target/targeting/common/entitypath.H:
```diff
@@ -0,0 +1,121 @@
+
+    uint16_t getSize() const
+    {
+        uint8_t size = _header & 0x0F;
```
uint16_t ?

gcpin commented on target/targeting/common/entitypath.C:
```diff
@@ -0,0 +1,188 @@
+    assert(getSize() < MAX_PATH_ELEMENTS);
+    PathElement pe;
+    pe.setType(type);
+    pe.instance = instance;
```
we can add explicit constructor to PathElement to initialize these instead ?

gcpin commented on target/targeting/common/entitypath.C:
```diff
@@ -0,0 +1,188 @@
+    if (type() != rhs.type())
+        return type() < rhs.type();
+
+    int cmp =
```
can be made const

gcpin commented on target/targeting/target_service.C:
```diff
@@ -0,0 +1,78 @@
+
+TargetPtr TargetService::getTopLevelTarget() const noexcept
+{
+    if (!_targetMap)
```
if we are checking for _targetMap for initialization everywhere , do we still need _initialized ?

gcpin commented on target/targeting/target.H:
```diff
@@ -0,0 +1,95 @@
+class Target;
+using TargetPtr = std::unique_ptr<Target>;
+class TargetDevtreeMap;
+class Target
```
we could rename this to phal_target / cec_target / hw_target  / host_target or something unique to phal team to make it easier to refer to like pdbg targets. target name is generic, and gets mixed up easily when we refer to it.

devenrao committed:
targeting: cache accessed targets

To support optional attributes that have application-level existence,
we now maintain a map of all queried targets.

When a target is requested:
1) If it was already accessed, a pointer to the existing instance is
returned.
2) Otherwise, a new Target is created, cached, and returned.

This commit lays the groundwork for associating hardware access objects
with targets that have the ATTR_DEVICE_PATH attribute set.

Tested:
1) Ensured existing unit test cases do not fail
2) Ensured cached target pointer is returned when queried for an
   cached target

[ RUN      ] TargetServiceTest.GetTopLevelTarget_IsCached
[       OK ] TargetServiceTest.GetTopLevelTarget_IsCached (0 ms)

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I42653475eb06e74709a0c1c309055b19b5e3b44c

gcpin commented on target/targeting/target_devtree_map.C:
```diff
@@ -54,67 +54,70 @@ TargetPtr TargetDevtreeMap::getParentOf(const TargetPtr& child,
-    const Target* source, AssociationType type, RecursionLevel recursionLevel,
-    const PredicateBase* predicate) const
+ConstTargetPtrList TargetDevtreeMap::getAssociated(
+    const ConstTargetPtr source, AssociationType type,
```
ConstTargetPtr source ?

devenrao committed:
targeting: Add getAttr, trySetAttr, tryGetAttr methods to Target

- Extended the Target class with generic getAttr, trySetAttr, and
  tryGetAttr  methods for attribute access.
- Implemented support for AttributeTraits::TypeStdArr (e.g.,
  std::array<char, N>)
  in get/set attribute logic.
- Enhanced setAttrImpl and tryGetAttrImpl with special handling for
  std::array<char, N> attributes such as ATTR_LOCATION_CODE.

This improves type safety and simplifies attribute access using C++
templates.

Tested:
[       OK ] TargetServiceTest.TestGetAttrThrowsOnMissing (0 ms)
[ RUN      ] TargetServiceTest.TestSetAttrThrowsOnMissing
[       OK ] TargetServiceTest.TestSetAttrThrowsOnMissing (0 ms)
[ RUN      ] TargetServiceTest.TestGetAttrAndTrySetAttr_Type
[       OK ] TargetServiceTest.TestGetAttrAndTrySetAttr_Type (0 ms)
[ RUN      ] TargetServiceTest.TestGetAttrAndTrySetAttr_LOCATION_CODE
[       OK ] TargetServiceTest.TestGetAttrAndTrySetAttr_LOCATION_CODE (0 ms)

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I6ae770e9c4bcf8ea974ef6bebff8f9cce1b26b54

devenrao committed:
targeting: support app-scoped optional attributes

Add support for optional attributes that exist only within the context
of the application and are not backed by the device tree. Specifically,
the optional attribute ATTR_HWACCESS_METHOD is now supported for
TYPE_PROC targets.

These optional attributes are:
Stored locally within the application
Not serialized into the device tree
Isolated per application (not shared across contexts)

A GMock-based test verifies that getAttr() prefers the optional attribute
when present and avoids accessing the device tree path.

Tested:
Edit
[       OK ] TargetServiceTest.getAttrPrefersOptionalOverFdt (0 ms)
[----------] 17 tests from TargetServiceTest (1 ms total)

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I3fb59afa9e3c010f5d677376095fac9ebedf76e3

devenrao committed:
targeting: fix code review comments

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I40109b660c75646636f7a1670f5a3bd6dd844a68
